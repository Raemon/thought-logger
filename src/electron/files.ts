import path from "path";
import fs from "node:fs/promises";
import { format, parse, setDefaultOptions, startOfWeek } from "date-fns";
import { app } from "electron";
import {
  Keylog,
  Screenshot,
  Summary,
  SummaryScopeTypes,
} from "../types/files.d";
import { Dirent } from "node:fs";
import log from "../logging";

setDefaultOptions({ weekStartsOn: 1 });

const userDataPath = app.getPath("userData");

function groupByWeek<T>(record: Record<string, T>): Map<string, T[]> {
  const groups = new Map<string, T[]>();

  for (let dateString of Object.keys(record)) {
    const date = parse(dateString, "yyyy-MM-dd", new Date());
    const week = format(startOfWeek(date), "YYYY-'W'ww", {
      useAdditionalWeekYearTokens: true,
    });

    if (groups.has(week)) {
      const groupData = groups.get(week);
      groupData.push(record[dateString]);
    } else {
      groups.set(week, [record[dateString]]);
    }
  }

  return groups;
}

async function readFilesFromDirectory(path: string): Promise<Dirent[]> {
  try {
    await fs.access(path);
  } catch (error) {
    log.error(`Couldn't access ${path}: ${error}`);
    return [];
  }

  return fs
    .readdir(path, {
      recursive: true,
      withFileTypes: true,
    })
    .then((files) => files.filter((file) => file.isFile()));
}

async function maybeReadContents(path: string): Promise<string | null> {
  try {
    await fs.access(path);
  } catch (error) {
    log.info(`No file at ${path}.`);
    return null;
  }

  return fs.readFile(path, { encoding: "utf-8" });
}

export async function getRecentSummaries(): Promise<Summary[]> {
  const keylogsPath = path.join(userDataPath, "files", "keylogs");
  const screenshotsPath = path.join(userDataPath, "files", "screenshots");

  const dailySummaries: Record<string, Summary> = {};
  const keylogs: Record<string, Keylog> = {};
  const screenshots: Record<string, Record<string, Screenshot>> = {};

  const keylogFiles = await readFilesFromDirectory(keylogsPath);

  for (let file of keylogFiles) {
    const fileName = path.basename(file.name);
    const dir = file.parentPath;
    const result = fileName.match(/[^.]+/);
    const dateString = result[0];
    let date = parse(dateString, "yyyy-MM-dd", new Date());

    keylogs[dateString] = keylogs[dateString] || {
      appPath: path.join(dir, dateString + ".processed.by-app.log"),
      chronoPath: path.join(dir, dateString + ".processed.chronological.log"),
      rawPath: path.join(dir, dateString + ".log"),
      date,
    };
  }

  const screenshotFiles = await readFilesFromDirectory(screenshotsPath);

  for (let file of screenshotFiles) {
    const fileName = path.basename(file.name);
    const dir = file.parentPath;
    const result = fileName.match(/[^.]+/);
    const dateString = result[0];
    let date = parse(dateString, "yyyy-MM-dd HH_mm_ss", new Date());
    const dayString = format(date, "yyyy-MM-dd");

    screenshots[dayString] = screenshots[dayString] || {};
    screenshots[dayString][dateString] = {
      imagePath: path.join(dir, dateString + ".jpg"),
      summaryPath: path.join(dir, dateString + ".txt"),
      date,
    };
  }

  const dateStrings = Array.from(
    new Set(Object.keys(keylogs).concat(Object.keys(screenshots))),
  );

  for (let dateString of dateStrings) {
    const availableKeylogs = keylogs[dateString]
      ? [keylogs[dateString]]
      : ([] as Keylog[]);
    const availableScreenshots = screenshots[dateString]
      ? Object.values(screenshots[dateString])
      : ([] as Screenshot[]);

    const date = parse(dateString, "yyyy-MM-dd", new Date());
    const monthString = format(date, "yyyy-MM");

    const summaryPath = path.join(
      userDataPath,
      "files",
      "keylogs",
      monthString,
      `${dateString}.aisummary.log`,
    );

    const contents = await maybeReadContents(summaryPath);

    dailySummaries[dateString] = dailySummaries[dateString] || {
      path: summaryPath,
      contents,
      date,
      keylogs: availableKeylogs,
      screenshots: availableScreenshots,
      loading: false,
      scope: SummaryScopeTypes.Day,
    };
  }

  const weeklyKeylogs = groupByWeek(keylogs);
  const weeklyScreenshots = groupByWeek(screenshots);

  const weeks: string[] = Array.from(
    new Set(
      weeklyKeylogs.keys().toArray().concat(weeklyScreenshots.keys().toArray()),
    ),
  );

  const weeklySummaries: Summary[] = [];

  for (let week of weeks) {
    const date = parse(week, "YYYY-'W'ww", new Date(), {
      useAdditionalWeekYearTokens: true,
    });
    const summaryPath = path.join(
      userDataPath,
      "files",
      `${week}.aisummary.log`,
    );
    const contents = await maybeReadContents(summaryPath);

    weeklySummaries.push({
      path: summaryPath,
      contents,
      date,
      keylogs: weeklyKeylogs.get(week)
        ? weeklyKeylogs.get(week)
        : ([] as Keylog[]),
      loading: false,
      scope: SummaryScopeTypes.Week,
      screenshots: weeklyScreenshots.get(week)
        ? weeklyScreenshots
            .get(week)
            .reduce(
              (acc, screenshot) => acc.concat(Object.values(screenshot)),
              [],
            )
        : ([] as Screenshot[]),
    });
  }

  return weeklySummaries.concat(Object.values(dailySummaries));
}

const TWO_WEEKS_IN_SECONDS = 60 * 60 * 24 * 7;

async function walkDir(dir: string, allEntries: string[]) {
  const entries = await fs.readdir(dir, { withFileTypes: true });
  for (const entry of entries) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      await walkDir(full, allEntries);
    } else {
      allEntries.push(full);
    }
  }
}

export async function recentFiles(
  ageInSeconds: number = TWO_WEEKS_IN_SECONDS,
): Promise<string[]> {
  const filesDir = path.join(userDataPath, "files");
  try {
    const allEntries: string[] = [];

    await walkDir(filesDir, allEntries);

    // Sort by modification time descending
    const datedPaths: { path: string; mtime: number }[] = [];
    for (const filePath of allEntries) {
      // Skip .DS_Store files
      if (path.basename(filePath) === ".DS_Store") continue;
      // Skip screenshots
      if (path.extname(filePath) === ".jpg") continue;

      const stat = await fs.stat(filePath);
      datedPaths.push({ path: filePath, mtime: stat.mtimeMs });
    }
    datedPaths.sort((a, b) => b.mtime - a.mtime);

    // Return a limited list, e.g. 20 items
    const nowMs = Date.now();
    return datedPaths
      .filter((x) => nowMs - x.mtime <= ageInSeconds * 1000)
      .map((x) => x.path);
  } catch (error) {
    log.error("Failed to list recent files:", error);
    return [];
  }
}
