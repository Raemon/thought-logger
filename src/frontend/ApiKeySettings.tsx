import React, { useEffect, useState } from "react";

const ApiKeySettings = () => {
  const [apiKey, setApiKey] = useState<string>("");
  const [keyStatus, setKeyStatus] = useState<{hasKey: boolean, message: string} | null>(null);
  const [isSaving, setIsSaving] = useState<boolean>(false);
  const [message, setMessage] = useState<{text: string, isError: boolean} | null>(null);

  useEffect(() => {
    checkApiKey();
  }, []);

  const checkApiKey = async () => {
    const status = await window.openRouter.checkApiKey();
    setKeyStatus(status);
  };

  const saveApiKey = async () => {
    if (!apiKey.trim()) {
      setMessage({ text: "API key cannot be empty", isError: true });
      return;
    }

    setIsSaving(true);
    setMessage(null);

    try {
      const result = await window.openRouter.saveApiKey(apiKey);
      
      if (result.success) {
        setMessage({ text: result.message, isError: false });
        setApiKey("");
        checkApiKey();
      } else {
        setMessage({ text: result.message, isError: true });
      }
    } catch (error) {
      setMessage({ text: "An error occurred while saving the API key", isError: true });
    } finally {
      setIsSaving(false);
    }
  };

  return (
    <div>
      <h3>OpenRouter API Key</h3>
      
      <div style={{ marginBottom: 12 }}>
        {keyStatus && (
          <div style={{ color: keyStatus.hasKey ? "#0a0" : "#a00" }}>
            {keyStatus.message}
          </div>
        )}
      </div>
      
      <div style={{ marginBottom: 16 }}>
        <input
          type="password"
          value={apiKey}
          onChange={(e) => setApiKey(e.target.value)}
          placeholder="Enter OpenRouter API key"
          style={{ width: "300px" }}
        />
        <button 
          onClick={saveApiKey} 
          disabled={isSaving}
          style={{ marginLeft: 8 }}
        >
          {isSaving ? "Saving..." : "Save"}
        </button>
      </div>
      
      {message && (
        <div style={{ color: message.isError ? "#a00" : "#0a0" }}>
          {message.text}
        </div>
      )}
      
      <div style={{ fontSize: 12, marginTop: 12 }}>
        <em>
          OpenRouter API key is required for text extraction from screenshots.
          Get your API key at <a 
            href="#" 
            onClick={(e) => {
              e.preventDefault();
              window.userData.openExternalUrl("https://openrouter.ai/keys");
            }}
            style={{ color: "blue", textDecoration: "none", cursor: "pointer" }}
          >openrouter.ai/keys</a>
        </em>
      </div>
    </div>
  );
};

export default ApiKeySettings; 