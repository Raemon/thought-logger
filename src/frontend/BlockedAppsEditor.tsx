import React, { useEffect, useState } from "react";

const BlockedAppsEditor = () => {
  const [blockedApps, setBlockedApps] = useState<string[]>([]);
  const [newApp, setNewApp] = useState<string>("");

  useEffect(() => {
    window.preferences.getPreferences().then((prefs) => {
      setBlockedApps(prefs.blockedApps);
    });
  }, []);

  const addApp = async () => {
    if (newApp.trim() !== "") {
      const updatedApps = [...blockedApps, newApp.trim()];
      setBlockedApps(updatedApps);
      await window.preferences.setPreferences({ blockedApps: updatedApps });

      setNewApp("");
    }
  };

  const removeApp = async (appToRemove: string) => {
    const updatedApps = blockedApps.filter((app) => app !== appToRemove);
    setBlockedApps(updatedApps);
    await window.preferences.setPreferences({ blockedApps: updatedApps });
  };

  return (
    <div className="p-5">
      <h3 className="text-xl mb-2.5">Blocked Apps</h3>
      <table style={{ borderCollapse: "collapse", marginBottom: 16 }}>
        <tbody>
          {blockedApps
            .sort()
            .reverse()
            .map((app) => (
              <tr key={app}>
                <td>{app}</td>
                <td style={{ textAlign: "right" }}>
                  <button
                    className="border-2 border-red-400 hover:bg-red-400 hover:text-white text-red-400 font-bold rounded ml-2 px-2 py-0.5 text-sm"
                    onClick={() => removeApp(app)}
                  >
                    Remove
                  </button>
                </td>
              </tr>
            ))}
        </tbody>
      </table>
      <input
        type="text"
        value={newApp}
        onChange={(e) => setNewApp(e.target.value)}
        className="mb-2.5 p-2 border-2 rounded"
        placeholder="Add new app"
      />
      <button
        onClick={addApp}
        className="bg-blue-500 hover:bg-blue-700 text-white font-bold rounded ml-2 px-2 py-0.5"
      >
        Add
      </button>
    </div>
  );
};

export default BlockedAppsEditor;
