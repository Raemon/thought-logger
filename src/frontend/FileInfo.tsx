import React, { useEffect, useState, useCallback } from "react";
import { groupBy } from "lodash";

function openFolder() {
  window.userData.openUserDataFolder();
}

const isAiSummary = (file: string) => file.includes("aisummary");

const getFileName = (file: string) => {
  const parts = file.split("/").slice(-1)[0].split(".");
  if (isAiSummary(file)) return "ai summary";
  return parts.slice(1, 3).join(".");
};

function getMonth(dateStr: string) {
  const d = new Date(dateStr);
  return d.toLocaleString("en-US", { month: "long", year: "numeric" });
}

function getWeek(dateStr: string) {
  const d = new Date(dateStr);
  // Get Monday of the week
  const monday = new Date(d);
  monday.setDate(d.getDate() - ((d.getDay() + 6) % 7));
  return `${monday.getFullYear()}-W${String(getWeekNumber(monday)).padStart(2, "0")}`;
}

function getWeekNumber(d: Date) {
  // Copy date so don't modify original
  d = new Date(Date.UTC(d.getFullYear(), d.getMonth(), d.getDate()));
  // Set to nearest Thursday: current date + 4 - current day number
  d.setUTCDate(d.getUTCDate() + 4 - (d.getUTCDay() || 7));
  // Get first day of year
  const yearStart = new Date(Date.UTC(d.getUTCFullYear(), 0, 1));
  // Calculate full weeks to nearest Thursday
  const weekNo = Math.ceil(
    (((d as any) - (yearStart as any)) / 86400000 + 1) / 7,
  );
  return weekNo;
}

export function FileInfo() {
  const [dataFolder, setDataFolder] = useState("(loading)");
  const [filesByDate, setFilesByDate] = useState<Record<string, string[]>>({});
  const [fileContents, setFileContents] = useState<Record<string, string>>({});

  useEffect(() => {
    window.userData.getUserDataFolder().then(setDataFolder);
    window.userData.listRecentFiles().then((files) => {
      // Filter files from the past year
      const oneYearAgo = new Date();
      oneYearAgo.setFullYear(oneYearAgo.getFullYear() - 1);
      const filteredFiles = files.filter((file) => {
        const dateStr = file.split("/").slice(-1)[0].split(".")[0];
        const fileDate = new Date(dateStr);
        return !isNaN(fileDate.getTime()) && fileDate >= oneYearAgo;
      });
      setFilesByDate(
        groupBy(
          filteredFiles,
          (file) => file.split("/").slice(-1)[0].split(".")[0],
        ),
      );
    });
  }, []);

  useEffect(() => {
    const allFiles = Object.values(filesByDate).flat();
    Promise.all(
      allFiles.map((file) =>
        file in fileContents
          ? Promise.resolve()
          : window.userData
              .readFile(file)
              .then((content) =>
                setFileContents((prev) => ({ ...prev, [file]: content })),
              )
              .catch(() =>
                setFileContents((prev) => ({
                  ...prev,
                  [file]: "(error loading file)",
                })),
              ),
      ),
    );
  }, [filesByDate]);

  const getFormattedFile = (file: string) => {
    const content = fileContents[file];
    if (!content) return "Loading...";
    if (!content.match(/1\. /)) {
      return content;
    }
    const newContent = "1. " + content.split("1.")[1];
    const paragraphs = newContent.split(/\n\s*\n/);
    // Check if the last paragraph starts with a number
    const last = paragraphs[paragraphs.length - 1];
    if (last && !/^\d/.test(last.trim())) {
      paragraphs.pop();
    }
    return paragraphs.join("\n\n");
  };

  // Group dates by month and week
  const tocByMonth: Record<string, Record<string, string[]>> = {};
  Object.keys(filesByDate).forEach((dateStr) => {
    const month = getMonth(dateStr);
    const week = getWeek(dateStr);
    if (!tocByMonth[month]) tocByMonth[month] = {};
    if (!tocByMonth[month][week]) tocByMonth[month][week] = [];
    tocByMonth[month][week].push(dateStr);
  });

  return (
    <div>
      <h3>User files</h3>
      <p>
        This app takes screenshots at regular intervals (configurable below),
        and logs mouse and keyboard inputs.
      </p>
      <button onClick={openFolder} className="ml-0">
        Open Files Folder
      </button>
      <span className="ml-2 text-xs text-gray-500">{dataFolder}</span>

      <div className="flex gap-2.5">
        <a
          href="http://localhost:8765/today"
          onClick={(e) => {
            e.preventDefault();
            window.userData.openExternalUrl("http://localhost:8765/today");
          }}
          className="text-blue-600 no-underline cursor-pointer"
        >
          today
        </a>
        <a
          href="http://localhost:8765/yesterday"
          onClick={(e) => {
            e.preventDefault();
            window.userData.openExternalUrl("http://localhost:8765/yesterday");
          }}
          className="text-blue-600 no-underline cursor-pointer"
        >
          yesterday
        </a>
        <a
          href="http://localhost:8765/last7days"
          onClick={(e) => {
            e.preventDefault();
            window.userData.openExternalUrl("http://localhost:8765/week");
          }}
          className="text-blue-600 no-underline cursor-pointer"
        >
          week
        </a>
      </div>

      <div className="flex gap-5 mt-5">
        <div
          className="min-w-[220px] max-w-[250px] pr-2 border-r border-gray-200 text-sm"
          style={{ maxHeight: 600, overflowY: "auto" }}
        >
          <div className="font-bold mb-2">Table of Contents</div>
          {Object.entries(tocByMonth).map(([month, weeks]) => (
            <div key={month} className="mb-2">
              <div className="font-semibold">{month}</div>
              <div style={{ marginLeft: 10 }}>
                {Object.entries(weeks).map(([week, days]) => (
                  <div key={week} className="mb-1">
                    <div className="text-xs text-gray-600">
                      <a
                        href={`#week-${week}`}
                        className="text-blue-500 hover:underline"
                        onClick={(e) => {
                          e.preventDefault();
                          const el = document.getElementById(`week-${week}`);
                          if (el)
                            el.scrollIntoView({
                              behavior: "smooth",
                              block: "start",
                            });
                        }}
                      >
                        {week}
                      </a>
                    </div>
                    <div style={{ marginLeft: 10 }}>
                      {days
                        .sort((a, b) => b.localeCompare(a))
                        .map((day) => (
                          <div key={day}>
                            <a
                              href={`#day-${day}`}
                              className="text-blue-700 hover:underline"
                              style={{ fontSize: 13 }}
                              onClick={(e) => {
                                e.preventDefault();
                                const el = document.getElementById(
                                  `day-${day}`,
                                );
                                if (el)
                                  el.scrollIntoView({
                                    behavior: "smooth",
                                    block: "start",
                                  });
                              }}
                            >
                              {new Date(day).toLocaleDateString("en-US", {
                                weekday: "short",
                                day: "numeric",
                                month: "short",
                              })}
                            </a>
                          </div>
                        ))}
                    </div>
                  </div>
                ))}
              </div>
            </div>
          ))}
        </div>
        <div className="min-w-[300px] flex-1">
          {(() => {
            // Flatten and sort all dates descending
            const allDates = Object.keys(filesByDate).sort((a, b) =>
              b.localeCompare(a),
            );
            const result: React.ReactNode[] = [];
            allDates.forEach((date, idx) => {
              const files = filesByDate[date];
              const week = getWeek(date);
              // Render the day as before
              result.push(
                <div key={date} className="mb-2.5" id={`day-${date}`}>
                  <div className="mb-1 font-bold flex items-center">
                    <span className="mr-auto">
                      {new Date(date).toLocaleDateString("en-US", {
                        weekday: "long",
                        day: "numeric",
                        month: "long",
                        year: "numeric",
                      })}
                    </span>
                    <button
                      className="bg-blue-500 hover:bg-blue-700 text-white font-bold rounded ml-2 px-2 py-0.5 text-xs"
                      onClick={() => {
                        const rawFile = files.find(
                          (f) => !isAiSummary(f) && !f.includes(".processed."),
                        );
                        window.userData.generateAISummary(rawFile);
                      }}
                    >
                      regenerate summary
                    </button>
                    {files.find(
                      (f) => !isAiSummary(f) && !f.includes(".processed."),
                    ) && (
                      <button
                        className="ml-2 px-2 py-2 text-xs text-blue-800 font-normal"
                        onClick={() =>
                          window.userData.openExternalUrl(
                            `file://${files.find((f) => !isAiSummary(f) && !f.includes(".processed."))}`,
                          )
                        }
                      >
                        raw
                      </button>
                    )}
                    {files
                      .filter(
                        (f) => !isAiSummary(f) && f.includes(".processed."),
                      )
                      .map((f) => (
                        <button
                          key={f}
                          className="ml-2 px-2 py-0.5 text-xs text-green-800 font-normal"
                          onClick={() =>
                            window.userData.openExternalUrl(`file://${f}`)
                          }
                        >
                          {f.split(".").slice(1, 3).join(".")}
                        </button>
                      ))}
                  </div>
                  <div className="flex flex-col gap-1">
                    {files
                      .sort((a, b) =>
                        isAiSummary(a) === isAiSummary(b)
                          ? a.localeCompare(b)
                          : isAiSummary(a)
                            ? -1
                            : 1,
                      )
                      .map((file) => {
                        if (isAiSummary(file)) {
                          return (
                            <div
                              key={file}
                              className="whitespace-pre-wrap bg-gray-100 p-3 rounded text-sm"
                            >
                              {getFormattedFile(file)}
                            </div>
                          );
                        }
                      })}
                  </div>
                </div>,
              );
              // If this is the last day of the week (Sunday or last date or week changes), render the week summary
              const isSunday = new Date(date).getDay() === 0;
              const isLast = idx === allDates.length - 1;
              const nextWeek = !isLast ? getWeek(allDates[idx + 1]) : null;
              if (isSunday || isLast || nextWeek !== week) {
                result.push(
                  <div
                    key={`week-title-${week}`}
                    id={`week-${week}`}
                    className="font-semibold text-3xl text-base mt-4 mb-2 mt-10"
                  >
                    Week of{" "}
                    {(() => {
                      // Find the Monday of this week
                      const d = new Date(date);
                      d.setDate(d.getDate() - ((d.getDay() + 6) % 7));
                      return d.toLocaleDateString("en-US", {
                        month: "long",
                        day: "numeric",
                        year: "numeric",
                      });
                    })()}
                  </div>,
                );
              }
            });
            return result;
          })()}
        </div>
      </div>
    </div>
  );
}
