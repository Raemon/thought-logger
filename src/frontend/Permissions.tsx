import React, { useCallback, useEffect, useState } from "react";

function usePermissions() {
  const [perm, setPerm] = useState<Record<string, string> | null>(null);

  const refreshPermissions = useCallback(() => {
    window.permissions
      .requestPermissionsStatus()
      .then((permissions) => setPerm(permissions));
  }, []);

  useEffect(() => {
    refreshPermissions();
  }, []);

  return [perm, refreshPermissions] as const;
}

export function Permissions() {
  const [perm, refreshPermissions] = usePermissions();
  return (
    <div className="p-5">
      <h3 className="text-xl mb-2.5">Permissions</h3>
      {perm &&
        Object.entries(perm).map(([k, v]) => (
          <div key={k} style={{ textTransform: "capitalize" }}>
            {k}:{" "}
            {v === "granted" ? (
              <span style={{ color: "#0a0" }}>granted</span>
            ) : (
              <span style={{ color: "#a00" }}>denied</span>
            )}
          </div>
        ))}
      <div style={{ fontSize: 12, marginTop: 12, marginBottom: 12 }}>
        <em>
          Keylogging requires both "accessibility" and "input monitoring"
          permissions in the <b>System Preferences: Privacy & Security</b>.
        </em>
      </div>
      <button
        onClick={refreshPermissions}
        style={{ marginTop: 12 }}
        className="bg-blue-500 hover:bg-blue-700 text-white font-bold rounded px-2 py-0.5"
      >
        Refresh
      </button>
    </div>
  );
}
