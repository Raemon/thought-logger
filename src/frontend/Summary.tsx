import React, { ReactElement } from "react";
import { format } from "date-fns";
import { SerializedLog, SerializedScopeTypes } from "../types/files.d";

export default function Summary({ log }: { log: SerializedLog }): ReactElement {
  const date = log.date;
  const dateStr = format(date, "yyyy-MM-dd");
  return (
    <div key={dateStr} className="mb-2.5" id={`day-${dateStr}`}>
      <div className="mb-1 font-bold flex items-center">
        <span className="mr-auto">
          {log.scope === SerializedScopeTypes.Week && "Week of "}
          {date.toLocaleDateString("en-US", {
            weekday:
              log.scope === SerializedScopeTypes.Day ? "long" : undefined,
            day: "numeric",
            month: "long",
            year: "numeric",
          })}
        </span>
        <button
          className="bg-blue-500 hover:bg-blue-700 text-white font-bold rounded ml-2 px-2 py-0.5 text-xs"
          onClick={() => window.userData.generateAISummary(log)}
        >
          regenerate summary
        </button>
        {log.rawPath && (
          <button
            className="ml-2 px-2 py-2 text-xs text-blue-800 font-normal"
            onClick={() =>
              window.userData.openExternalUrl(`file://${log.rawPath}`)
            }
          >
            Raw
          </button>
        )}
        {log.appPath && (
          <button
            className="ml-2 px-2 py-0.5 text-xs text-green-800 font-normal"
            onClick={() =>
              window.userData.openExternalUrl(`file://${log.appPath}`)
            }
          >
            By App
          </button>
        )}
        {log.chronoPath && (
          <button
            className="ml-2 px-2 py-0.5 text-xs text-green-800 font-normal"
            onClick={() =>
              window.userData.openExternalUrl(`file://${log.chronoPath}`)
            }
          >
            Chronological
          </button>
        )}
      </div>
      {log.summaryContents && (
        <div className="flex flex-col gap-1">
          <div className="whitespace-pre-wrap bg-gray-100 p-3 rounded text-sm">
            {log.summaryContents}
          </div>
        </div>
      )}
    </div>
  );
}
