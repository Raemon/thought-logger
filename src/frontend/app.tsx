import React, { useState, useEffect } from "react";
import { FileInfo } from "./FileInfo";
import { Permissions } from "./Permissions";
import { ScreenshotController } from "./ScreenshotController";
import BlockedAppsEditor from "./BlockedAppsEditor";
import ApiKeySettings from "./ApiKeySettings";
import SummarySettings from "./SummarySettings";

const TabButton = ({
  selected,
  onClick,
  children,
}: {
  selected: boolean;
  onClick: () => void;
  children: React.ReactNode;
}) => (
  <button
    onClick={onClick}
    className={
      `px-4 py-2 border-none cursor-pointer mr-0.5 ` +
      (selected
        ? "bg-white border-b-0"
        : "bg-gray-100 border-b border-gray-300")
    }
  >
    {children}
  </button>
);

export function App() {
  const [activeTab, setActiveTab] = useState<"logs" | "settings">("settings");
  const [hasLogs, setHasLogs] = useState(false);

  useEffect(() => {
    // Check if there are any log files
    window.userData.getRecentLogs().then((logs) => {
      if (logs.length > 0) {
        setHasLogs(true);
        setActiveTab("logs");
      }
    });
  }, []);

  return (
    <div className="bg-white p-4">
      <div className="border-b border-gray-300 mb-4">
        <TabButton
          selected={activeTab === "logs"}
          onClick={() => setActiveTab("logs")}
        >
          Logs
        </TabButton>
        <TabButton
          selected={activeTab === "settings"}
          onClick={() => setActiveTab("settings")}
        >
          Settings
        </TabButton>
      </div>

      {activeTab === "logs" && <FileInfo />}

      {activeTab === "settings" && (
        <>
          <ScreenshotController />
          <SummarySettings />
          <ApiKeySettings />
          <Permissions />
          <BlockedAppsEditor />
        </>
      )}
    </div>
  );
}
