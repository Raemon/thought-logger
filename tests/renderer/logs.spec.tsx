import React from "react";
import { afterAll, beforeAll, expect, test, vi } from "vitest";
import { render } from "vitest-browser-react";

import { FileInfo } from "../../src/frontend/FileInfo";
import { SerializedScopeTypes } from "../../src/types/files.d";

const userData: UserData = {
  openUserDataFolder: vi.fn<UserData["openUserDataFolder"]>(),
  generateAISummary: vi
    .fn<UserData["generateAISummary"]>()
    .mockResolvedValue("Stub summary"),
  getRecentApps: vi.fn<UserData["getRecentApps"]>().mockResolvedValue(["foo"]),
  getRecentLogs: vi.fn<UserData["getRecentLogs"]>().mockResolvedValue([
    {
      date: new Date(2025, 8, 19),
      loading: false,
      scope: SerializedScopeTypes.Day,
      appPath: "fakepath.by-app.log",
      chronoPath: "fakepath.chronological.log",
      rawPath: "fakepath.log",
      summaryContents: "This is a daily summary",
    },
    {
      date: new Date(2025, 8, 12),
      loading: false,
      scope: SerializedScopeTypes.Week,
      appPath: "fakepath.by-app.log",
      chronoPath: "fakepath.chronological.log",
      rawPath: "fakepath.log",
      summaryContents: "This is a weekly summary",
    },
  ]),
  getUserDataFolder: vi
    .fn<UserData["getUserDataFolder"]>()
    .mockResolvedValue("/User/test/files"),
  onUpdateRecentLogs: vi.fn<UserData["onUpdateRecentLogs"]>(),
  openExternalUrl: vi.fn<UserData["openExternalUrl"]>(),
  openFile: vi.fn<UserData["openFile"]>(),
  readFile: vi
    .fn<UserData["readFile"]>()
    .mockResolvedValue("fake file contents"),
};

vi.stubGlobal("userData", userData);

beforeAll(() => {
  vi.useFakeTimers();
  const date = new Date(2025, 8, 19);
  vi.setSystemTime(date);
});

afterAll(() => {
  vi.useRealTimers();
});

test("Renders the logs", async () => {
  const { asFragment, getByText } = render(<FileInfo />);
  await expect.element(getByText("This is a daily summary")).toBeVisible();
  expect(asFragment()).toMatchSnapshot();
});
