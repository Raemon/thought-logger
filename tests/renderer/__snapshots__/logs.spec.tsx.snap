// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders the logs 1`] = `
<DocumentFragment>
  <div>
    <h3>
      User files
    </h3>
    <p>
      This app takes screenshots at regular intervals (configurable below), and logs mouse and keyboard inputs.
    </p>
    <button
      class="ml-0"
    >
      Open Files Folder
    </button>
    <span
      class="ml-2 text-xs text-gray-500"
    >
      /User/test/files
    </span>
    <div
      class="flex gap-2.5"
    >
      <a
        class="text-blue-600 no-underline cursor-pointer"
        href="http://localhost:8765/today"
      >
        today
      </a>
      <a
        class="text-blue-600 no-underline cursor-pointer"
        href="http://localhost:8765/yesterday"
      >
        yesterday
      </a>
      <a
        class="text-blue-600 no-underline cursor-pointer"
        href="http://localhost:8765/last7days"
      >
        week
      </a>
    </div>
    <div
      class="flex gap-5 mt-5"
    >
      <div
        class="min-w-[220px] max-w-[250px] pr-2 border-r border-gray-200 text-sm"
        style="max-height: 600px; overflow-y: auto;"
      >
        <div
          class="font-bold mb-2"
        >
          Table of Contents
        </div>
        <div
          class="mb-2"
        >
          <div
            class="font-semibold"
          >
            2025-09
          </div>
          <div
            style="margin-left: 10px;"
          >
            <div
              class="mb-1"
            >
              <div
                class="text-xs text-gray-600"
              >
                <a
                  class="text-blue-500 hover:underline"
                  href="#week-2025-W38"
                >
                  2025-W38
                </a>
              </div>
              <div
                style="margin-left: 10px;"
              >
                <div>
                  <a
                    class="text-blue-700 hover:underline"
                    href="#day-2025-09-19"
                    style="font-size: 13px;"
                  >
                    Thu, Sep 18
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="min-w-[300px] flex-1"
      >
        <div
          class="mb-2.5"
          id="day-2025-09-19"
        >
          <div
            class="mb-1 font-bold flex items-center"
          >
            <span
              class="mr-auto"
            >
              Friday, September 19, 2025
            </span>
            <button
              class="bg-blue-500 hover:bg-blue-700 text-white font-bold rounded ml-2 px-2 py-0.5 text-xs"
            >
              regenerate summary
            </button>
            <button
              class="ml-2 px-2 py-2 text-xs text-blue-800 font-normal"
            >
              Raw
            </button>
            <button
              class="ml-2 px-2 py-0.5 text-xs text-green-800 font-normal"
            >
              By App
            </button>
            <button
              class="ml-2 px-2 py-0.5 text-xs text-green-800 font-normal"
            >
              Chronological
            </button>
          </div>
          <div
            class="flex flex-col gap-1"
          >
            <div
              class="whitespace-pre-wrap p-3 rounded text-sm bg-gray-100"
            >
              This is a daily summary
            </div>
          </div>
        </div>
        <div
          class="mb-2.5"
          id="day-2025-09-12"
        >
          <div
            class="mb-1 font-bold flex items-center"
          >
            <span
              class="mr-auto"
            >
              Week of Monday, September 8, 2025 to Sunday, September 14, 2025
            </span>
            <button
              class="bg-blue-500 hover:bg-blue-700 text-white font-bold rounded ml-2 px-2 py-0.5 text-xs"
            >
              regenerate summary
            </button>
            <button
              class="ml-2 px-2 py-2 text-xs text-blue-800 font-normal"
            >
              Raw
            </button>
            <button
              class="ml-2 px-2 py-0.5 text-xs text-green-800 font-normal"
            >
              By App
            </button>
            <button
              class="ml-2 px-2 py-0.5 text-xs text-green-800 font-normal"
            >
              Chronological
            </button>
          </div>
          <div
            class="flex flex-col gap-1"
          >
            <div
              class="whitespace-pre-wrap p-3 rounded text-sm bg-sky-50"
            >
              This is a weekly summary
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
